pkg/nsx/services/builder.go

```diff
@@ -374,33 +408,24 @@ func (service *SecurityPolicyService) buildRuleAndGroups(obj *v1alpha1.SecurityP
 	ruleServiceEntries := service.buildRuleServiceEntries(&rule.Ports)
 	nsxRule.ServiceEntries = *ruleServiceEntries
 
-	if len(rule.AppliedTo) > 0 {
-		nsxRuleAppliedGroup, nsxRuleAppliedGroupPath, err = service.buildRuleAppliedGroup(obj, rule, ruleIdx)
-		if err == nil {
-			ruleGroups = append(ruleGroups, *nsxRuleAppliedGroup)
-		} else {
-			log.Error(err, "failed to build rule applied groups")
-			return nil, nil, err
-		}
-	} else {
-		if len(obj.Spec.AppliedTo) == 0 {
-			err = errors.New("appliedTo needs to be set in either spec or rules")
-			log.Error(err, "error while validating appliedTo field")
-			return nil, nil, err
-		} else if nsxRuleSrcGroupPath == "ANY" && nsxRuleDstGroupPath == "ANY" {
-			// NSX-T manager will report error if all the rule's scope/src/dst are "ANY".
-			// So if the rule's scope is empty while policy's not, the rule's scope also
-			// will be set to the policy's scope to avoid this case.
-			nsxRuleAppliedGroupPath = service.buildPolicyGroupPath(obj)
-		} else {
-			nsxRuleAppliedGroupPath = "ANY"
-		}
+	nsxRuleAppliedGroup, nsxRuleAppliedGroupPath, err = service.buildRuleAppliedToGroup(
+		obj,
+		rule,
+		ruleIdx,
+		nsxRuleSrcGroupPath,
+		nsxRuleDstGroupPath,
+	)
+	if err != nil {
+		return nil, nil, err
 	}
+	ruleGroups = append(ruleGroups, nsxRuleAppliedGroup)
 	nsxRule.Scope = []string{nsxRuleAppliedGroupPath}
 
-	log.V(1).Info("built rule and groups", "nsxRuleAppliedGroup", nsxRuleAppliedGroup, "nsxRuleSrcGroup", nsxRuleSrcGroup, "nsxRuleDstGroup", nsxRuleDstGroup, "action", *nsxRule.Action, "direction", *nsxRule.Direction)
+	log.V(1).Info("built rule and groups", "nsxRuleAppliedGroup", nsxRuleAppliedGroup,
+		"~", nsxRuleSrcGroup, "nsxRuleDstGroup", nsxRuleDstGroup,
```

Comment:
`"~"`?

Reply:
Thanks for bringing out the accident change in the previous patch.  I've changed the code to fix this error.